### PR TITLE
Add options for deleting models during training to save space

### PR DESCRIPTION
**Pull request by gcooper-isi**
Corresponds to https://github.com/gcooper-isi/DeepSpeedExamples/pull/15, https://github.com/gcooper-isi/DeepSpeedExamples/pull/21, and part of https://github.com/gcooper-isi/DeepSpeedExamples/pull/40/files

### CODE REVIEW

File: megatron/checkpointing.py [hunk 1]
```diff
@@ -89,7 +93,58 @@ def get_checkpoint_tracker_filename(checkpoints_path):
...
+            if new_model == old_model:
+                return
+        model_list.append((new_model, validation_loss))
+        save_model_list(model_list)
+
+
+def remove_old_models():
+    args = get_args()
+    if torch.distributed.get_rank() == 0 and args.only_keep_checkpoints is not None:
+        model_list = load_model_list(args)
```
Comment (wrichardson-isi) [inline megatron/checkpointing.py:129]: It looks like `args` may be extraneous.

Comment (gcooper-isi) [reply megatron/checkpointing.py]: Fixed

File: megatron/checkpointing.py [hunk 2]
```diff
@@ -89,7 +93,58 @@ def get_checkpoint_tracker_filename(checkpoints_path):
...
+                        (highest_loss is None or math.isnan(validation_loss) or validation_loss > highest_loss):
+                    worst_model_index = i
+                    highest_loss = validation_loss
+            model = model_list.pop(worst_model_index or 0)[0]
+            print("Deleting old model file: " + model)
+            try:
+                shutil.rmtree(model)
+            except FileNotFoundError:
+                pass
+        save_model_list(args, model_list)
```
Comment (wrichardson-isi) [inline megatron/checkpointing.py:144]: It looks like `args` may be extraneous here, too.  I forgot to click the "Add single comment" button earlier.

Comment (gcooper-isi) [reply megatron/checkpointing.py]: Fixed

File: megatron/training.py [hunk 3]
```diff
@@ -506,6 +505,11 @@ def evaluate_and_print_results(prefix, forward_step_func,
     print_rank_0(string)
     print_rank_0('-' * length)
 
+    args = get_args()
+    if args.loss_dict_key in total_loss_dict:
+        return total_loss_dict[args.loss_dict_key].item()
+    return None
```
Comment (wrichardson-isi) [inline megatron/training.py:511]: This is something of a minor quibble, but I did try specifying this argument with a bogus key.  The program simply proceeded along without notifying me.  If I am reading the code correctly, the validation loss values for all of the models would be `None`.  Perhaps a warning message here?

Comment (gcooper-isi) [reply megatron/training.py:514]: Done
